src/storage/mod.rs

```diff
@@ -1396,6 +1398,49 @@ impl<E: Engine, L: LockManager> Storage<E, L> {
                 .await?
         }
     }
+
+    /// Update max_ts with the start_ts and check the given range for locks
+    /// which blocks the reading.
+    pub fn check_memory_locks_in_ranges(
+        &self,
+        mut ctx: Context,
+        start_ts: TimeStamp,
+        ranges: Vec<KeyRange>,
+    ) -> impl Future<Output = Result<Option<LockInfo>>> {
+        let concurrency_manager = self.concurrency_manager.clone();
+        self.read_pool
+            .spawn_handle(
+                async move {
+                    concurrency_manager.update_max_ts(start_ts);
+                    let bypass_locks = TsSet::vec_from_u64s(ctx.take_resolved_locks());
+                    for range in &ranges {
```

Comment:
Is it possible that it needs to read single keys instead of ranges? Otherwise will it be represented by range [`key`, `key.append(0)`) ?

Reply:
If it needs to read single keys, it just passes ranges like in tidb_query.
It can be [`key`, `key.append(0)`). It can be a bit larger than it too. We just don't optimize point range like tidb_query does.